### ISSUE TITLE
Free style fuzzy finder for filenames

### ISSUE THREAD

**Issue by gfmm-q3d** [Aider-AI/aider#2156]
## Current Behavior

When adding files, a match of the exact literal that the user is typing is done against all repo files paths.

## Feature Request

It would ease the user experience to be able to look for files to add in a free style format, much like what FZF does.

### Example

We want to add a file named `my_file.py`. This feature would allow the user to find a match by typing e.g. `my file` or even `file my`,  also by matching parts of the full path.

### UI Suggestion

Separating the strings to search for by a blank space is currently not possible, because of how the `/add` command and arguments is parsed. We could enter a hypothetical `fzf` mode by by typing `TAB` after a blank while typing the args for command `/add`. Also for command `/read-only`.

**Comment by paul-gauthier** [comment]
This looks like a duplicate of #492. Please see the comments there for more information, and feel free to continue the discussion within that issue.

I'm going to close this issue for now. But please let me know if you think this is actually a distinct issue and I will reopen this issue.

**Comment by gfmm-q3d** [comment]
Thank, @paul-gauthier!! I've read the Issue and closed PR. I understand why it can't be added as of now. I hope it will be available in a standard and validated way some day, through plugins, etc.

**Comment by raine** [comment]
For tmux users, here's a much better way to find files within aider.

- It opens a tmux popup with fzf for selecting files within the **current directory** where aider is launched. This is better than built in autocomplete which autocompletes files from the whole repository always, regardless of subdirectory.
- You may select multiple files with <kbd>tab</kbd>

### demo

https://github.com/user-attachments/assets/d2c0aeb9-3054-4579-9d12-fa2309f2f305

### the script

Put this somewhere in your path. I have mine in `~/bin`.

On mac you need to install the following with homebrew: `brew install coreutils fd`

```sh
#!/usr/bin/env bash

pane_dir=$(tmux display-message -p '#{pane_current_path}')
pane_id=$(tmux display-message -p '#{pane_id}')
cd "$pane_dir" || exit
git_root=$(git rev-parse --show-toplevel)

# gnu coreutils realpath is needed to make the paths relative to the currently active
# tmux pane relative to the git repo root, because aider always wants paths relative
# to the repo root, even if you are in a subdirectory
if [[ "$OSTYPE" == "darwin"* ]]; then
	if ! command -v grealpath >/dev/null 2>&1; then
		echo "grealpath not found. Install with: brew install coreutils" >&2
		exit 1
	fi
	realpath_cmd="grealpath"
else
	realpath_cmd="realpath"
fi

selected_files=$(
	fd --type f |
		fzf --multi \
			--reverse \
			--preview 'bat --style=numbers --color=always {}' |
		while read -r file; do
			"$realpath_cmd" --relative-to="$git_root" "$pane_dir/$file"
		done
)

if [ -n "$selected_files" ]; then
	files_oneline=$(echo "$selected_files" | tr '\n' ' ' | sed 's/ $//')
	tmux send-keys -t "$pane_id" "$files_oneline"
fi

```

### usage

Bind to the key of your preference. This binds it to <kbd>Prefix</kbd> + <kbd>Ctrl-f</kbd>.

```tmux
# ~/.tmux.conf
bind C-f display-popup -E "pick-files"
```

https://github.com/raine/tmux-file-picker

**Comment by gfmm-q3d** [comment]
Awesome, thanks @raine! I'll try it out.

**Comment by ZeekoZhu** [comment]
My fzf+[kitty](https://github.com/kovidgoyal/kitty) solution:

*~/.config/kitty/kitty_file_selector.fish*
```fish
#!/usr/bin/fish

set TARGET_WINDOW_ID $argv[1]

# Get git root and determine realpath command
set git_root (git rev-parse --show-toplevel)

# Select files using fzf
set selected_files (fd --type f | fzf --multi --reverse --preview 'bat --style=numbers --color=always {}'  --delimiter '-,/' --bind=ctrl-p:toggle,tab:ignore)

# map to relative path to git root with realpath
set selected_files (for file in $selected_files; realpath --relative-to=$git_root $file; end)

# Send selected files to kitty
if test -n "$selected_files"
    set files_oneline (string join ' ' $selected_files)
    kitty @ send-text --match id:$TARGET_WINDOW_ID $files_oneline
end
```

*~/.config/kitty/kitty.conf*
```
# pick files
map ctrl+shift+p launch --type overlay --cwd current --copy-env ~/.config/kitty/kitty_file_selector.fish @active-kitty-window-id
```

**Comment by raine** [comment]
Sharing one more useful script I've been using for quite some time now.

It allows picking files with fzf before starting aider. I have it named just `a` so I usually just run `a` in my project directory and pick the files I need.

```sh
#!/usr/bin/env bash

set -euo pipefail

selected_files=$(
	fzf \
		--multi \
		--preview 'bat --style=numbers --color=always {} 2>/dev/null || cat {}' \
		--preview-window='right:60%:wrap' \
		--bind='ctrl-/:change-preview-window(hidden|)' \
		--bind='esc:cancel' \
		--header="Select files with TAB, confirm with ENTER"
)

if [ -z "$selected_files" ]; then
	echo "No files selected. Exiting..."
	exit 0
fi

# Convert newline-separated files into space-separated for aider
files_for_aider=$(echo "$selected_files" | tr '\n' ' ')
aider $files_for_aider "$@"
```

**Comment by dmitrym0** [comment]
Any suggestions to make fzf available during aider sessions that's terminal agnostic? Tmux version looks beautiful though.
